flang/lib/Lower/Bridge.cpp

```diff
@@ -1903,26 +1903,26 @@ class FirConverter : public Fortran::lower::AbstractConverter {
                   isNumericScalarCategory(lhsType->category());
               auto rhs = isNumericScalar ? genExprValue(assign.rhs, stmtCtx)
                                          : genExprAddr(assign.rhs, stmtCtx);
-              auto lowerAllocatableLHS = [&]() -> fir::ExtendedValue {
-                auto lhs = genExprMutableBox(loc, assign.lhs);
-                llvm::SmallVector<mlir::Value> lengthParams;
-                if (auto *charBox = rhs.getCharBox())
-                  lengthParams.push_back(charBox->getLen());
-                else if (lhs.isDerivedWithLengthParameters())
-                  TODO(loc, "assignment to derived type allocatable with "
-                            "length parameters");
-                fir::factory::genReallocIfNeeded(
-                    *builder, loc, lhs, /*lbounds=*/llvm::None,
-                    /*shape=*/llvm::None, lengthParams);
-                // Assume lhs is not polymorphic for now given TODO above,
-                // otherwise, the read would is conservative and returns
-                // BoxValue for derived types.
-                return fir::factory::genMutableBoxRead(
-                    *builder, loc, lhs, /*mayBePolymorphic=*/false);
-              };
-              auto lhs = isWholeAllocatable(assign.lhs)
-                             ? lowerAllocatableLHS()
-                             : genExprAddr(assign.lhs, stmtCtx);
+              bool lhsIsWholeAllocatable = isWholeAllocatable(assign.lhs);
+              llvm::Optional<fir::factory::MutableBoxReallocation> lhsRealloc;
+              llvm::Optional<fir::MutableBoxValue> lhsMutableBox;
+              auto lhs = [&]() -> fir::ExtendedValue {
+                if (lhsIsWholeAllocatable) {
+                  lhsMutableBox = genExprMutableBox(loc, assign.lhs);
+                  llvm::SmallVector<mlir::Value> lengthParams;
+                  if (auto *charBox = rhs.getCharBox())
+                    lengthParams.push_back(charBox->getLen());
+                  else if (lhsMutableBox->isDerivedWithLengthParameters())
```

Comment:
The first test is on the source but the else test is on the destination. Is that correct?